File: src/org/mtransit/android/commons/provider/AvailabilityPercentStatusFilter.java
package org.mtransit.android.commons.provider;

import org.json.JSONException;
import org.json.JSONObject;
import org.mtransit.android.commons.MTLog;
import org.mtransit.android.commons.data.POI;

public class AvailabilityPercentStatusFilter extends StatusFilter {

	private static final String TAG = AvailabilityPercentStatusFilter.class.getSimpleName();

	@Override
	public String getLogTag() {
		return TAG;
	}

	public AvailabilityPercentStatusFilter(String targetUUID) {
		super(POI.ITEM_STATUS_TYPE_AVAILABILITY_PERCENT, targetUUID);
	}

	@Override
	public StatusFilter fromJSONStringStatic(String jsonString) {
		return fromJSONString(jsonString);
	}

	public static StatusFilter fromJSONString(String jsonString) {
		try {
			return fromJSON(new JSONObject(jsonString));
		} catch (JSONException jsone) {
			MTLog.w(TAG, jsone, "Error while parsing JSON string '%s'", jsonString);
			return null;
		}
	}

	public static StatusFilter fromJSON(JSONObject json) {
		try {
			String targetUUID = StatusFilter.getTargetUUIDFromJSON(json);
			AvailabilityPercentStatusFilter availabilityPercentStatusFilter = new AvailabilityPercentStatusFilter(targetUUID);
			StatusFilter.fromJSON(availabilityPercentStatusFilter, json);
			return availabilityPercentStatusFilter;
		} catch (JSONException jsone) {
			MTLog.w(TAG, jsone, "Error while parsing JSON object '%s'", json);
			return null;
		}
	}

	@Override
	public String toJSONStringStatic(StatusFilter statusFilter) {
		return toJSONString(statusFilter);
	}

	private static String toJSONString(StatusFilter statusFilter) {
		try {
			return toJSON(statusFilter).toString();
		} catch (JSONException jsone) {
			MTLog.w(TAG, jsone, "Error while generating JSON string '%s'", statusFilter);
			return null;
		}
	}

	private static JSONObject toJSON(StatusFilter statusFilter) throws JSONException {
		try {
			JSONObject json = new JSONObject();
			StatusFilter.toJSON(statusFilter, json);
			// if (statusFilter instanceof AvailabilityPercentStatusFilter) {
			// }
			return json;
		} catch (JSONException jsone) {
			MTLog.w(TAG, jsone, "Error while parsing JSON object '%s'", statusFilter);
			return null;
		}
	}

}

File: src/main/java/org/mtransit/android/commons/data/AvailabilityPercentStatusFilter.kt
package org.mtransit.android.commons.data

import org.json.JSONException
import org.json.JSONObject
import org.mtransit.android.commons.MTLog
import org.mtransit.android.commons.provider.status.StatusProviderContract

data class AvailabilityPercentStatusFilter(
    override val cacheOnly: Boolean? = null,
    override val cacheValidityInMs: Long? = null,
    override val inFocus: Boolean? = null,
    override val providedEncryptKeysMap: Map<String, String>? = null,
    override val targetUUID: String,
) : StatusProviderContract.Filter(POI.ITEM_STATUS_TYPE_AVAILABILITY_PERCENT, targetUUID) {

    companion object {
        private val LOG_TAG: String = AvailabilityPercentStatusFilter::class.java.getSimpleName()

        @JvmStatic
        fun from(poi: POI) = from(poi.uuid)

        @JvmStatic
        fun from(targetUUID: String) = AvailabilityPercentStatusFilter(
            targetUUID = targetUUID,
        )

        @JvmStatic
        fun fromJSONString(jsonString: String?): StatusProviderContract.Filter? {
            try {
                return jsonString?.let { fromJSON(JSONObject(it)) }
            } catch (jsone: JSONException) {
                MTLog.w(LOG_TAG, jsone, "Error while parsing JSON string '$jsonString'")
                return null
            }
        }

        fun fromJSON(json: JSONObject): StatusProviderContract.Filter? {
            try {
                return AvailabilityPercentStatusFilter(
                    cacheOnly = getCacheOnlyFromJSON(json),
                    cacheValidityInMs = getCacheValidityInMsFromJSON(json),
                    inFocus = getInFocusFromJSON(json),
                    providedEncryptKeysMap = getProvidedEncryptKeysMapFromJSON(json),
                    targetUUID = getTargetUUIDFromJSON(json),
                )
            } catch (jsone: JSONException) {
                MTLog.w(LOG_TAG, jsone, "Error while parsing JSON object '$json'")
                return null
            }
        }

        private fun toJSONString(statusFilter: StatusProviderContract.Filter) = toJSON(statusFilter)?.toString()

        private fun toJSON(statusFilter: StatusProviderContract.Filter): JSONObject? {
            try {
                return JSONObject().apply {
                    toJSON(statusFilter, this)
                }
            } catch (jsone: JSONException) {
                MTLog.w(LOG_TAG, jsone, "Error while making JSON object '$statusFilter'")
                return null
            }
        }
    }

    override fun getLogTag() = LOG_TAG

    override fun copyWith(providedEncryptKeysMap: Map<String, String>?) = this.copy(providedEncryptKeysMap = providedEncryptKeysMap)

    override fun fromJSONStringStatic(jsonString: String?) = fromJSONString(jsonString)

    override fun toJSONString() = toJSONString(this)

    override fun toJSONStringStatic(statusFilter: StatusProviderContract.Filter) = toJSONString(statusFilter)
}
